pages/contact/forms.py

```diff
@@ -0,0 +1,181 @@
+"""Contact form definitions and anti-spam validation.
+
+This module defines the `ContactForm` used by the contact page. It includes
+layered, user-transparent anti-spam checks (honeypot, submission timing, and
+double-submit cookie) and content validation.
+
+The form accepts `request` in its constructor to access cookies for the
+double-submit verification.
+"""
+
+from __future__ import annotations
+
+import re
+import time
+from typing import Any, Optional
+
+from django import forms
+from django.core import signing
+from django.core.exceptions import ValidationError
+
+
+URL_REGEX = re.compile(r"(https?://|www\.)", re.IGNORECASE)
+
+# Validation and anti-spam constants
+MIN_SUBMIT_SECONDS = 2
+MAX_TOKEN_AGE_SECONDS = 60 * 60
+MAX_URLS_ALLOWED = 3
+
+
+class ContactForm(forms.Form):
+	"""Contact form with layered anti-spam.
+
+	Fields:
+		name: Optional full name of the sender (2–100 chars when provided).
+		email: Optional reply address, validated when provided.
+		message: Main text body, 20–5000 chars, with URL and HTML limits.
+		suggestion, dm_support, other: Category checkboxes, at least one must
+		    be selected.
+		website: Honeypot field, must remain empty.
+		contact_ts: Signed timestamp token to check submission timing.
+		contact_dsc: Token that must match the cookie for double-submit check.
+
+	Anti-spam strategy:
+		- Honeypot rejects naive bots filling hidden fields.
+		- TimestampSigner token rejects submissions that are too fast (<2s) or
+		  too old (>60 minutes).
+		- Double-submit cookie reduces scripted posts and replays.
+	"""
+
+	name = forms.CharField(min_length=2, max_length=100, required=False)
+	email = forms.EmailField(required=False)
+	message = forms.CharField(min_length=20, max_length=5000, widget=forms.Textarea)
+	suggestion = forms.BooleanField(required=False)
+	dm_support = forms.BooleanField(
+		required=False, label="Request for data management/sharing support"
+	)
+	other = forms.BooleanField(required=False)
+
+	# Anti-spam fields
+	website = forms.CharField(required=False, widget=forms.HiddenInput)
+	contact_ts = forms.CharField(widget=forms.HiddenInput, strip=False)
+	contact_dsc = forms.CharField(widget=forms.HiddenInput, strip=False)
+
+	# Internal state for logging (not exposed to users)
+	_blocked_reason: Optional[str] = None
+
+	def __init__(self, *args: Any, request=None, **kwargs: Any) -> None:
+		"""Initialise the form.
+
+		Args:
+			*args: Positional form args.
+			request: Optional HttpRequest to access cookies for token checks.
+			**kwargs: Keyword form args.
+		"""
+		super().__init__(*args, **kwargs)
+		self.request = request
+
+	def clean_email(self) -> str:
+		"""Validate email and prevent header injection.
+
+		Returns:
+			The sanitised email value.
+
+		Raises:
+			ValidationError: If CR/LF characters are present.
+		"""
+		value = self.cleaned_data.get("email", "")
+		if "\r" in value or "\n" in value:
+			raise ValidationError("Enter a valid email address.")
```

Comment:
Is it possible to have `\r` or `\n` in the email field ? I wasn't able to enter it in the form when I tired.

Reply:
Yes, because browsers typically prevent it as a default built-in security measure.
But this extra validation layer is still beneficial to prevent header injection via `curl` POSTs requests for example

Reply:
Ah ok! now I see the point of it 😅